WIP - security

e-schultz opened the pull request:
@yuri @igor-ka @SethDavenport @bennett000 @winkerVSbecks Still very much WIP - just wanting to get some early eyes on this. Will be fleshing out more of it on Fri/Mon/Tues 

Pulling over some of our general security guide-lines, and going to start fleshing out some of the angular specific sections next.

igor-ka commented:
@e-schultz thoroughly enjoyed reading this. Really nice summary.
:+1:

SethDavenport commented:
:+1:

I'll have to play around more with lucsa.  Helmet is already enabled in the starters.

e-schultz commented:
@igor-ka @SethDavenport still working on fleshing out this section. the JWT revoke/expire is just in point form right now

bennett000 commented on handout/security/cookies-and-tokens.md:
```diff
@@ -0,0 +1,44 @@
+
+## Cookies
+
+The other option to storing JWT information in Cookies, this however comes with it's own set of security challenges, such as still being vulnerable to CSRF attacks. If storing your JWT in a Cookie, it is a good idea to have them be HTTPS only to ensure they are not transmitted across an insecure connection. However, when making a Cookie HTTPS only - your JavaScript application can no longer read  ...
```
_"it's own set of security..." should be "its"

I do like the `/user/whoami` technique that gets around the `localStorage` issue, although it seems like unless you have a dedicated login server domain you end up sending the cookie with every request (even though the server only pays attention to whoami)

bennett000 commented:
Is it worth mentioning rate limiting node servers?

e-schultz commented:
@bennett000 any recommended rate limiting packages?

bennett000 commented:
@e-schultz good question.  None of the first four express solutions I looked at look amazing, but 

[express-limiter]:(https://www.npmjs.com/package/express-limiter) or [strict-rate-limiter]:(https://www.npmjs.com/package/strict-rate-limiter) work in parallel and have some basic options.

I was kind of hoping for something more comprehensive with white/black lists as well as a tar pit function.

e-schultz commented:
@igor-ka @SethDavenport If get time, I want to try and re-write some of the security code examples to the ngcourse-next style (was adapted from some existing code I have), but will be merging this in later today. If someone could take a quick read over that'd be great

e-schultz commented:
also want to add a bit about controllers/error handling - maybe pull in a bit of the controller factory / error handling. But won't get a chance to do that until tonight

SethDavenport commented on handout/security/angular-http.md:
```diff
@@ -0,0 +1,29 @@
+
+However, Angular does provide the ability to help prevent against cross-site request forgery - CSRF, or sometimes referred to as XSRF. If you are using the [lusca](https://github.com/krakenjs/lusca) middleware that was described in the [node modules](./node-modules.md) section, it will set the appropriate defaults for the header and cookie settings.
+
+This will create an JavaScript readable `XSRF-TOKEN` cookie on the first HTTP GET request, and this will then be sent as an `X-XSRF-TOKEN` header with subsequent requests.
```
not sure about lucsa, but some CSRF solutions actually update the token on each request.

SethDavenport commented on handout/security/angular-sce.md:
````diff
@@ -0,0 +1,23 @@
+<div ng-bind-html="userInput"></div>
+```
+
+And it will display the formatted input. In newer versions of Angular, it will throw an error unless you have included the `ngSanatize` module, and the most current version of the `angular-sanatize` script for your application.
````
Typo: ngSanitize, angular-sanitize

SethDavenport commented on handout/security/node-modules.md:
```diff
@@ -0,0 +1,46 @@
+# Node Modules
+
+Keeping in mind with the principals of _Avoid DIY Security_ and _Choose third-party software wisely._, there are two commonly used node modules that can be used to help secure your node servers - [helmet](https://github.com/helmetjs/helmet), and [lusca](https://github.com/krakenjs/lusca) by PayPal.  Both of these offer similar, but different sets of functionality.
```
typo: principles

SethDavenport commented on handout/security/node-modules.md:
```diff
@@ -0,0 +1,46 @@
+
+### Lucsa and CSRF
+
+One of the noteworthy features of Lucasa, and their CSRF support, is an Angular specific that will et lusca up to use the default settings for CSRF validation according to the [AngularJS docs](https://docs.angularjs.org/api/ng/service/$http#cross-site-request-forgery-xsrf-protection).
```
There's some overlap between the two libs.  In cases where they both do something (e.g. frameguard, HSTS), what's the better alternative?

SethDavenport commented on handout/security/principals/authentication.md:
```diff
@@ -0,0 +1,206 @@
+
+In rare cases you will need to be able to retrieve the original password. In
+those cases you will need to encrypt it and worry about how you manage the
+keys.
```
Is it worth mentioning timing attacks here? Or does bcrypt solve that? Or perhaps only certain implementations solve it?

Is there a go-to implementation of bcrypt that we trust?

odziem replied:
I suggest the go-to bcrypt library should be the "bcrypt" npm module, which is safe against timing attacks if you use it as advertised. See [StackOverflow](http://stackoverflow.com/questions/35620979/is-bcrypt-compare-vulnerable-to-timing-attack).

SethDavenport commented on handout/security/principals/node.md:
```diff
@@ -0,0 +1,89 @@
+
+## 2. Use a tool to block common attacks.
+
+NPM's module [helmetjs](https://www.npmjs.com/package/helmet) provides
```
This point is a bit redundant since it was discussed in more detail above.

SethDavenport commented:
Looks really good overall.  Just a couple typos and optional comments.

e-schultz commented:
cleaned up what I could,  + also re added some of the missing flux handouts to the summary, and adjusted the code examples to flow better so when copy/pasting aspects of it the person running the course shouldn't run into issues.
